### DISCUSSION TITLE
상품정보 수정시 태그 단어로 인한 실패 문제

### DISCUSSION THREAD

**Discussion by importunate-dev** [commerce-api-naver/commerce-solutions#50] [묻고 답하기] [answered]
안녕하세요. 저희 서비스는 고객의 상품 상세페이지를 수정해주는 기능을 활용중입니다. 간혹, 
"태그로 등록할 수 없는 단어가 포함되었습니다. "
라는 메시지와 함께 수정 요청이 실패하는 경우가 발생하고 있습니다. 상품 태그는 전혀 수정하지 않고 조회 요청에서 받은 데이터 그대로에서 detailcontent만 수정했음에도 이런 메시지가 나와 의문입니다.

1. 애초에 등록할 수 없는 태그가 왜, 어떻게 등록되어 있는지 모르겠습니다. 혹시 상품 태그에 관한 제한사항이 생기기 전에 등록된 태그인 경우 발생할 수 있는 문제일까요?
2. 혹시나 싶어 상품 수정시 조회로 받은 데이터에서 detailAttibute 내에 있는 seoInfo 값만 삭제하여 요청을 보냈더니, 기존 상품 태그가 모두 삭제되는 문제가 발생했습니다. 상품태그와 무관하게 상세페이지만 조회할 수 있는 별도의 방법은 없을까요?
3. 상품태그 이외에도 이와 같은 문제를 일으킬 여지가 있는 상세페이지 작성 정책에 대해 정리된 문서가 있는지, 없다면 메시지를 통해 개발사 쪽에서 해결해야하는 문제인지 여쭤보고 싶습니다.

**Comment by commerce-api-naver** [comment, marked as answer]
안녕하세요. 커머스API 이민석입니다.

### 1. 상품 태그 등록 관련
말씀해주신 부분은 스마트스토어 상품 데이터의 설계 구조와 관련 정책이 유기적으로 동작하는 부분이기에 일괄적으로 설명하긴 어렵습니다.
주로 발생하는 케이스는 아래와 같은 케이스들입니다.
- (언급해주신 것과 같이) 금지 태그 설정 이전에 상품에 등록된 태그
- 상품의 카테고리 이동으로 금지 태그 내용의 변동
- 그 밖에 매우 드문 기술적 문제

문제가 발생할 당시 대상 호출건의 Trace ID 등을 문의주시면 케이스별로 확인이 가능합니다.
좀 더 살펴보았으면 하는 케이스가 있으시다면 대상 호출건의 오류 메시지 Response body 전문 혹은 Trace ID를 알려주시면 확인 도와드리겠습니다.

### 2. 상품상세설명 내용만 취급할 수 있는 별도의 방법
> (문의는 상품상세설명의 '조회'를 문의주셨으나 맥락상 '수정/변경'을 의도하신 것으로 보여 '수정/변경'의 의미로 답변드립니다)

현재로서는 상품 정보 수정 목적으로 `PATCH` Method API 형태는 제공하지 않습니다.
스마트스토어 상품 데이터들 중 일부 정보는 상호간의 유기적으로 동작하는 경우가 많습니다. 따라서 수정 과정에서도 (스마트스토어센터와 마찬가지로) 대상 상품의 전체 정보를 수정하는 방식으로 API를 제공하고 있습니다. (`PUT` Method API)
그렇기에 전체 상품 정보 중 특정한 필드를 제외하고는 Request body에 포함하지 않을 경우 저장된 정보를 제거하는 의도로 해석하며, 상품상세설명 내용만 수정할 수 있는 기능은 현재 제공하지 않습니다.

### 3. 상품 수정 시 유효성 검증과 관련한 개발사 대응
현 시점에서는 Response body 내 반환 메시지를 통한 대응을 권해드립니다.

스마트스토어는 다양한 유형의 상품군을 취급할 수 있습니다.
다양한 유형의 상품군에 적용되는 각종 현행법률을 준수하고 관련 기관으로부터의 권고사항을 이행하며 구매자 분들께 최선의 구매 경험을 드릴 수 있도록 판매자 분들이 입력하는 상품 정보는 각 카테고리별로 다양한 유효성 검증 정책을 적용하게 됩니다.
이러한 유효성 검증 정책은 카테고리별로 사정이 매우 다르거니와 그 규모도 방대하여 개별 솔루션 설계에 녹아내기 어렵습니다.

따라서 상품 정보 수정 시 내부 정책에 위반되거나 충족하지 못한 수정 요청건은 보통 API 호출에 따른 Response body 내 오류 메시지 내용으로 구체적인 사유를 반환합니다.
이 내용을 최종 사용자(판매자)에게 그대로 표시하거나 솔루션 내부에서 처리하는 방향으로 설계를 권해드리고 있습니다.

**Reply by importunate-dev** [reply]
친절한 답변 감사합니다. 개발에 많은 도움이 될 것 같습니다.
